README.md
# ProportionalMultiBw

JBrowse plugin to display multiple BigWig tracks as a stacked proportion of depth.

[![Build Status](https://travis-ci.org/cancerit/proportionalmultibw.svg?branch=master)](https://travis-ci.org/cancerit/proportionalmultibw) : master
[![Build Status](https://travis-ci.org/cancerit/proportionalmultibw.svg?branch=develop)](https://travis-ci.org/cancerit/proportionalmultibw) : develop

`ProportionalMultiBw` is based on [MultiBigWig](https://github.com/elsiklab/multibigwig).  Significant credit and thanks are due to [Colin Diesh](https://github.com/cmdcolin) for help and guidance.

## Install

Unpack the release into the plugins directory of JBrowse.  The folder must be
named:

```
$JBROWSE_ROOT/plugins/ProportionalMultiBw
```

After unpack you should be able to see:

```
$JBROWSE_ROOT/plugins/ProportionalMultiBw/img/zoom_in.png
```

(sorry, limitations of finding the resource, get in touch if you have a solution)

## Associated software

Data files for use in this tool can be generated using the C package [cgpBigWig](https://github.com/cancerit/cgpBigWig).

## Available plots

### ProportionalXYPlot

`ProportionalXYPlot` provides a track able to read a set of BigWig files to produce a stacked proportion bar plot for each position.

This expects 2 or more BigWig files:

1. BigWig containing counts of total data at a position, e.g.
    * Total read depth
    * Patients screened
2. 1 or more BigWig files of fractional data (sum of a position should equal 1), e.g.
    * Allele fration (4 bw files, A, C, G, T)
    * Patients `+ve`/`-ve` for a kmown somatic mutation

The `bw` files for fractional data should not contain `0` values.

Using this set of files an allele stack is rendered proportional to the allele fractions (full height of the track) and the depth overlaid.

#### Configuration - Allele fraction

Allele stacks are only rendered if the position matches reference below the default threshold of `0.9`, this is configurable both in base configuration and via a track menu option.

Stacks can also be configured to hide when a minimum reference allele fraction is passed.

![ProportionalXYPlot native scale](img/XY.png)

```js
{
  // standard Wiggle options
  "storeClass"     : "ProportionalMultiBw/Store/SeqFeature/ProportionalMultiBw",
  "type"           : "ProportionalMultiBw/View/Track/ProportionalWiggle/ProportionalXYPlot",
  "label"          : "ProportionalMultiBw-XY",
  "scale"          : "linear",
  "yScalePosition" : "right",
  // content of ACGT urlTemplates matched against ref
  "matchRef"       : true,
  // Stacks hidden if reference allele exceeds this fraction
  "maxRefFrac"     : 0.9,
  // Stacks hidden if reference allele below this fraction
  "minRefFrac"     : 0.0,
  // Optional labels to apply to mouseovers, falls back to 'name'
  "labels" : {
    "counts"       : "Depth"
  },
  "urlTemplates"   : [
    // Data for raw count, e.g. total sequencing depth
    // MUST be present
     { "url"  : "bw/SAMPLE.bam.bw",
       "name"  : "counts",
       "color" : "black"
     },
     // Data files for allele fractions
     { "url"   : "bw/A.SAMPLE.bam.bw",
       "name"  : "A",
       "color" : "#00BF00"
     },
     { "url"   : "bw/C.SAMPLE.bam.bw",
       "name"  : "C",
       "color" : "#4747ff"
     },
     { "url"   : "bw/G.SAMPLE.bam.bw",
       "name"  : "G",
       "color" : "#d5bb04"
     },
     { "url"   : "bw/T.SAMPLE.bam.bw",
       "name"  : "T",
       "color" : "#f00"
     }
  ]
}
```

For high depth regions the log scale is preferable:

![ProportionalXYPlot log scale](img/log.png)

```js
{
  ...
  "scale": "log",
  ...
}
```

By default only positions with less than 90% match to reference are shown.  A track menu item allows this to be modified:

![ProportionalXYPlot max fraction dialoge](img/maxFrac.png)

E.g. modified to 0.92:

![ProportionalXYPlot max fraction 0.92](img/maxFrac92pct.png)

Zooming out to less than 1px per b.p. results in the fraction stacks being disabled:

![ProportionalXYPlot less than 1 px per b.p.](img/lt1pxPerBp.png)

#### Configuration - Non-ref matched stacks

A minor change to the configuration allows stacks to be generated based on presence of data alone rather than how they diverge from reference alleles.  This can be used for population statistics such as comparing the number of patients presenting with a SNP vs. the number screened.

Changes to the config would be along the lines:

```js
{
  ...
  // Disable matching of non 'count' data against reference.
  "matchRef" : false,
  // Data files for allele fractions
  "urlTemplates" : [
    { "url"   : "bw/donors_screened.bw",
      "name"  : "counts",
      "color" : "black"
    },
    { "url"   : "bw/pos.bw",
      "name"  : "pos",
      "color" : "#00BF00"
    },
    { "url"   : "bw/neg.bw",
      "name"  : "neg",
      "color" : "#4747ff"
    },
  ],
  "labels" : {
    "counts" : "Screened",
    "pos"    : "+ve",
    "neg"    : "-ve",
  }
  ...
}
```

### ProportionalDensity

At present this has not been modified from the original base taken from [https://github.com/elsiklab/multibigwig](MultiBigWig).  The intent is to give a produce a single density plot showing hotspots of divergence from reference.

--

## Why?

Over the last few years it has been common to generate whole genome sequencing (WGS) data
around 30-50x deep.  Even standard whole exome sequencing (WXS) is rarely over 100x and
JBrowse can display these reasonably well.

Within our group we are more regularly carrying out amplicon sequencing experiments (targeted pulldown)
with depth in the 1000s causing visualisation of the profile of errors/mutations in a region difficult.

Even in WGS, due to the new XTen Illumina machines, more high depth experiments are being run to detect
improve sensitivity to subclonal events in cancer genomes.


## Preparing a release

### Testing

Testing is performed using jasmine:

As per `.travis.yml` install the necessary tools

```
npm install -g eslint eslint-plugin-react babel-eslint bower phantomjs-prebuilt
pip install --user RangeHTTPServer
bower install
```

Start a simple web-server:

```
python -m RangeHTTPServer&
```

Run the tests:

```
phantomjs test/run-jasmine.js http://localhost:8000/test/
```

Don't forget to kill off the server (`fg` then `Ctrl-C`)

### Release

* Ensure any changes to documentation are complete.
* Bump the version number in `js/main.js`.
* Edit CHANGES.md.
* Push to github and with for tests to pass.
  * Fix if tests fail.
  * Version of JBrowse is hard coded in `bower.json` due to master branch being volatile.
* Raise merge request or finish hotfix.


CHANGES.md
# CHANGES

## 1.3.1

Fix undefined variable checking

## 1.3.0

Updates to work with webpack version of JBrowse (1.13+).

## 1.2.0

Adds min reference fraction track option.

## 1.1.0

... ignore ...

## 1.0.4

Unconfirmed change in JBrowse core resulted in functional method to determine
location of plugin install being inoperable.  Now requires plugin to be installed
in `plugins/ProportionalMultiBw`.

## 1.0.3

Bug fixes and minor option cleanup

* Occasional display of allele stack for one allele when no depth
* Flipping to global autoscale when changing max ref proportion
* Cleaned up menu removing the global/local autoscale options.

## 1.0.2

Large improvement in performance

* Easily view counts data across a 250Mbases.
* Resolves #3: Defaults `chunkSizeLimit`.
* Resolves #4: Some doc fixes.

## 1.0.1

Correctes bug in region jumping via search box, method of holding config was to blame

## 1.0.0

First release


test/spec/ProportionalMultiBw.spec.js
require([
    'dojo/_base/declare',
    'dojo/_base/array',
    'JBrowse/Browser',
    'ProportionalMultiBw/View/Track/ProportionalWiggle/ProportionalXYPlot',
    'ProportionalMultiBw/Store/SeqFeature/ProportionalMultiBw'
],
function(
    declare,
    array,
    Browser,
    ProportionalXYPlot,
    ProportionalMultiBwStore
) {

    describe('Can initialize track:', function() {

        var track = new ProportionalXYPlot({
            browser: new Browser({unitTestMode: true}),
            config: {
                labels: {"counts": "Depth"},
                counts: { "name": "counts","color": "black","urlTemplate": "volvox-proportionalMultiBw.bw" },
                urlTemplates: [
                {"color": "#00BF00","url": "A.volvox-proportionalMultiBw.bw","name": "A"},
                {"color": "#4747ff","url": "C.volvox-proportionalMultiBw.bw","name": "C"},
                {"color": "#d5bb04","url": "G.volvox-proportionalMultiBw.bw","name": "G"},
                {"color": "#f00","url": "T.volvox-proportionalMultiBw.bw","name": "T"},
                ],
                label: "testtrack"
            }
        });
        it('track', function() {
            expect(track).toBeTruthy();
        });

    });

    describe('Can initialize store:', function() {

        var store = new ProportionalMultiBwStore({
            browser: new Browser({unitTestMode: true}),
            urlTemplates: [
                {"color": "black",   "url": "../data/counts.volvox-proportionalMultiBw.bw",   "name": "counts"},
                {"color": "#00BF00", "url": "../data/A.volvox-proportionalMultiBw.bw", "name": "A"},
                {"color": "#4747ff", "url": "../data/C.volvox-proportionalMultiBw.bw", "name": "C"},
                {"color": "#d5bb04", "url": "../data/G.volvox-proportionalMultiBw.bw", "name": "G"},
                {"color": "#f00",    "url": "../data/T.volvox-proportionalMultiBw.bw", "name": "T"},
            ]
        });
        // set required browser variables specific to ProportionalMultiBw
        store.browser.view = {pxPerBp: 1};

        it('store', function() {
            expect(store).toBeTruthy();
        });

    });

    describe('Checking pxPerBp >=1 counts:', function() {

        var store = new ProportionalMultiBwStore({
            browser: new Browser({unitTestMode: true}),
            urlTemplates: [
                {"color": "black",   "url": "../data/counts.volvox-proportionalMultiBw.bw",   "name": "counts"},
                {"color": "#00BF00", "url": "../data/A.volvox-proportionalMultiBw.bw", "name": "A"},
                {"color": "#4747ff", "url": "../data/C.volvox-proportionalMultiBw.bw", "name": "C"},
                {"color": "#d5bb04", "url": "../data/G.volvox-proportionalMultiBw.bw", "name": "G"},
                {"color": "#f00",    "url": "../data/T.volvox-proportionalMultiBw.bw", "name": "T"},
            ]
        });
        // set required browser variables specific to ProportionalMultiBw
        store.browser.view = {pxPerBp: 1};

        var features = [];
        beforeEach(function(done) {
          features = [];
          store.getFeatures({ref: "ctgA", start: 1, end: 100}, function(feature) {
            features.push(feature);
          }, function() {
            done();
          }, function() {
            console.error(error);
            done();
          });
        });

        // test values in files with:
        // bwcat -i some.bw -r ctgA:1-101 | grep -vcE '\t0.000000'
        it('get bigwig values - A', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="A"; })
          expect(bwarr.length).toEqual(25); // result is ctg:1-101 if testing on cmd
        });
        it('get bigwig values - C', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="C"; })
          expect(bwarr.length).toEqual(18);
        });
        it('get bigwig values - G', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="G"; })
          expect(bwarr.length).toEqual(24);
        });
        it('get bigwig values - T', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="T"; })
          expect(bwarr.length).toEqual(30);
        });
        it('get bigwig values - Counts', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="counts"; })
          expect(bwarr.length).toEqual(13);
        });


    });

    describe('Checking pxPerBp <1 counts:', function() {

        var store = new ProportionalMultiBwStore({
            browser: new Browser({unitTestMode: true}),
            urlTemplates: [
                {"color": "black",   "url": "../data/counts.volvox-proportionalMultiBw.bw",   "name": "counts"},
                {"color": "#00BF00", "url": "../data/A.volvox-proportionalMultiBw.bw", "name": "A"},
                {"color": "#4747ff", "url": "../data/C.volvox-proportionalMultiBw.bw", "name": "C"},
                {"color": "#d5bb04", "url": "../data/G.volvox-proportionalMultiBw.bw", "name": "G"},
                {"color": "#f00",    "url": "../data/T.volvox-proportionalMultiBw.bw", "name": "T"},
            ]
        });
        // set required browser variables specific to ProportionalMultiBw
        store.browser.view = {pxPerBp: 0.5};

        var features = [];
        beforeEach(function(done) {
          features = [];
          store.getFeatures({ref: "ctgA", start: 1, end: 100}, function(feature) {
            features.push(feature);
          }, function() {
            done();
          }, function() {
            console.error(error);
            done();
          });
        });

        // test values in files with:
        // bwcat -i some.bw -r ctgA:1-101 | grep -vcE '\t0.000000'
        it('get bigwig values - A', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="A"; })
          expect(bwarr.length).toEqual(0); // result is ctg:1-101 if testing on cmd
        });
        it('get bigwig values - C', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="C"; })
          expect(bwarr.length).toEqual(0);
        });
        it('get bigwig values - G', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="G"; })
          expect(bwarr.length).toEqual(0);
        });
        it('get bigwig values - T', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="T"; })
          expect(bwarr.length).toEqual(0);
        });
        it('get bigwig values - Counts', function() {
          var bwarr = array.filter(features, function(f) { return f.get('score')>0&&f.get('source')=="counts"; })
          expect(bwarr.length).toEqual(13);
        });


    });

});



js/ProportionalMultiBw.profile.js
function copyOnly(mid) {
  return mid in {
    // There are no modules right now that are copy-only. If you have some, though, just add
    // them here like this:
    // 'app/module': 1
  };
}

var profile = {
  action: 'release',
  cssOptimize: 'comments',
  mini: true,

  basePath: '../../../src',
  packages: [
    {name: 'ProportionalMultiBw', location: '../plugins/ProportionalMultiBw/js' }
  ],

  layerOptimize: 'closure',
  stripConsole: 'normal',
  selectorEngine: 'acme',

  layers: {
    'ProportionalMultiBw/main': {
      include: [
        'ProportionalMultiBw'
      ],
      exclude: ['JBrowse']
    }
  },

  staticHasFeatures: {
    'dojo-trace-api': 0,
    'dojo-log-api': 0,
    'dojo-publish-privates': 0,
    'dojo-sync-loader': 0,
    'dojo-xhr-factory': 0,
    'dojo-test-sniff': 0
  },

  resourceTags: {
    // Files that contain test code.
    test: function(/* filename, mid */) {
      return false;
    },

    // Files that should be copied as-is without being modified by the build system.
    copyOnly: function(filename, mid) {
      return copyOnly(mid);
    },

    // Files that are AMD modules.
    amd: function(filename, mid) {
      return !copyOnly(mid) && /.js$/.test(filename);
    },

    // Files that should not be copied when the “mini” compiler flag is set to true.
    miniExclude: function(filename, mid) {
      return !(/^ProportionalMultiBw/.test(mid));
    }
  }
};


js/View/Dialog/MaxRefFracDialog.js
define([
  'dojo/_base/declare',
  'dojo/dom-construct',
  'dojo/on',
  'dijit/focus',
  'dijit/form/NumberSpinner',
  'dijit/form/Button',
  'JBrowse/View/Dialog/WithActionBar'
],
function(
  declare,
  dom,
  on,
  focus,
  NumberSpinner,
  Button,
  ActionBarDialog
) {
  return declare(ActionBarDialog, {
    title: 'Set max reference fraction',

    constructor: function(args) {
      this.maxRefFrac = args.maxRefFrac || 0.95;
      this.browser     = args.browser;
      this.setCallback   = args.setCallback || function() {};
      this.cancelCallback  = args.cancelCallback || function() {};
    },

    _fillActionBar: function(actionBar) {
      new Button({
        label: 'OK',
        onClick: dojo.hitch(this, function() {
          var height = +this.maxScoreSpinner.getValue();
          if (isNaN(height)) {
            return;
          }
          this.setCallback && this.setCallback(height);
          this.hide();
        })
      }).placeAt(actionBar);

      new Button({
        label: 'Cancel',
        onClick: dojo.hitch(this, function() {
          this.cancelCallback && this.cancelCallback();
          this.hide();
        })
      }).placeAt(actionBar);
    },

    show: function(/* callback */) {
      dojo.addClass(this.domNode, 'maxRefFracDialog');

      this.maxScoreSpinner = new NumberSpinner({
        value: this.maxRefFrac,
        smallDelta: 0.01,
        constraints: {min:0, max:1}
      });

      this.set('content', [
        dom.create('label', { 'for': 'read_depth', innerHTML: '' }),
        this.maxScoreSpinner.domNode,
        dom.create('span', { innerHTML: ' max ref frac' })
      ]);

      this.inherited(arguments);
    },

    hide: function() {
      this.inherited(arguments);
      window.setTimeout(dojo.hitch(this, 'destroyRecursive'), 500);
    }
  });
});


js/View/Dialog/MinRefFracDialog.js
define([
  'dojo/_base/declare',
  'dojo/dom-construct',
  'dojo/on',
  'dijit/focus',
  'dijit/form/NumberSpinner',
  'dijit/form/Button',
  'JBrowse/View/Dialog/WithActionBar'
],
function(
  declare,
  dom,
  on,
  focus,
  NumberSpinner,
  Button,
  ActionBarDialog
) {
  return declare(ActionBarDialog, {
    title: 'Set min reference fraction',

    constructor: function(args) {
      this.minRefFrac = args.minRefFrac || 0.0;
      this.browser     = args.browser;
      this.setCallback   = args.setCallback || function() {};
      this.cancelCallback  = args.cancelCallback || function() {};
    },

    _fillActionBar: function(actionBar) {
      new Button({
        label: 'OK',
        onClick: dojo.hitch(this, function() {
          var height = +this.minScoreSpinner.getValue();
          if (isNaN(height)) {
            return;
          }
          this.setCallback && this.setCallback(height);
          this.hide();
        })
      }).placeAt(actionBar);

      new Button({
        label: 'Cancel',
        onClick: dojo.hitch(this, function() {
          this.cancelCallback && this.cancelCallback();
          this.hide();
        })
      }).placeAt(actionBar);
    },

    show: function(/* callback */) {
      dojo.addClass(this.domNode, 'minRefFracDialog');

      this.minScoreSpinner = new NumberSpinner({
        value: this.minRefFrac,
        smallDelta: 0.01,
        constraints: {min:0, max:1}
      });

      this.set('content', [
        dom.create('label', { 'for': 'min_ref_frac', innerHTML: '' }),
        this.minScoreSpinner.domNode,
        dom.create('span', { innerHTML: ' min ref frac' })
      ]);

      this.inherited(arguments);
    },

    hide: function() {
      this.inherited(arguments);
      window.setTimeout(dojo.hitch(this, 'destroyRecursive'), 500);
    }
  });
});


js/View/Track/ProportionalWiggleBase.js
define([
  'dojo/_base/declare',
  'dojo/_base/array',
  'dojo/_base/lang',
  'dojo/_base/event',
  'dojo/dom-construct',
  'dojo/on',
  'dojo/mouse',
  'JBrowse/View/Track/WiggleBase',
  'ProportionalMultiBw/View/Dialog/MaxRefFracDialog',
  'ProportionalMultiBw/View/Dialog/MinRefFracDialog'
],
function(
  declare,
  array,
  lang,
  domEvent,
  dom,
  on,
  mouse,
  WiggleBase,
  MaxRefFracDialog,
  MinRefFracDialog
) {
  return declare(WiggleBase, {

    constructor: function(args) {
      this.labels = args.config.urlTemplates;
      if(args.config.hasOwnProperty('counts')) {
        this.labels.push(args.config.counts);
      }
    },

    fillBlock: function(args) {
      var thisB = this;
      this.heightUpdate(this._canvasHeight(), args.blockIndex);

      // hook updateGraphs onto the end of the block feature fetch
      var oldFinish = args.finishCallback || function() {};
      args.finishCallback = function() {
        thisB.updateGraphs(args, oldFinish);
      };

      // get the features for this block, and then set in motion the
      // updating of the graphs
      this._getBlockFeatures(args);
    },

    _calculatePixelScores: function(canvasWidth, features, featureRects) {

      var nameMap = {};
      for (var k = 0; k < this.labels.length; k++) {
        nameMap[this.labels[k].name] = k;
      }
      var pxPerBp = this.browser.view.pxPerBp;
      var pixelValues = new Array(canvasWidth);
      array.forEach(features, function(f, i) {
        if(pxPerBp < 1 && f.get('source') !== 'counts') return;
        var k = nameMap[f.get('source')];
        var fRect = featureRects[i];
        var jEnd = fRect.r;
        var score = f.get('score');
        for (var j = Math.round(fRect.l); j < jEnd; j++) {
          if (!pixelValues[j]) {
            pixelValues[j] = new Array(this.labels.length);
          }
          if (!pixelValues[j][k]) {
            pixelValues[j][k] = { score: score, feat: f };
          }
        }
      }, this);

      return pixelValues;
    },
    _trackMenuOptions: function() {
      var options = this.inherited(arguments);
      var track = this;
      options.push({
        label: 'Set min reference fraction',
        onClick: function() {
          new MinRefFracDialog({
            setCallback: function(filterInt) {
              track.config.minRefFrac = filterInt;
              track.browser.publish('/jbrowse/v1/c/tracks/replace', [track.config]);
            },
            minRefFrac: track.config.minRefFrac || 0.0
          }).show();
        }
      });
      options.push({
        label: 'Set max reference fraction',
        onClick: function() {
          new MaxRefFracDialog({
            setCallback: function(filterInt) {
              track.config.maxRefFrac = filterInt;
              track.browser.publish('/jbrowse/v1/c/tracks/replace', [track.config]);
            },
            maxRefFrac: track.config.maxRefFrac || 0.9
          }).show();
        }
      });

      return options;
    },
    _trackDetailsContent: function() {
      var ret = '';
      if (this.config.colorizeAbout) {
        array.forEach(this.labels, function(elt) {
          ret += '<div style="display: block; clear:both;"><div class="colorsquare" style="background: ' + elt.color + '"></div>' + elt.name;
        }, this);
      } else {
        ret = this.inherited(arguments);
      }
      return ret;
    },
    mouseover: function(bpX, evt) {
      if(bpX === undefined) {
        var thisB = this;
        thisB.scoreDisplay.flag.style.display = 'none';
        thisB.scoreDisplay.pole.style.display = 'none';
      }
      else {
        var block;
        array.some(this.blocks, function(b) {
          if(b && b.startBase <= bpX && b.endBase >= bpX) {
            block = b;
            return true;
          }
          return false;
        });

        if(!(block && block.canvas && block.pixelScores && evt)) return;

        var pixelValues = block.pixelScores;
        var canvas = block.canvas;
        var cPos = dojo.position(canvas);
        var cx = evt.pageX - cPos.x;

        if(this._showPixelValue(this.scoreDisplay.flag, pixelValues[ Math.round(cx) ])) {
          this.scoreDisplay.flag.style.display = 'block';
          this.scoreDisplay.pole.style.display = 'block';

          this.scoreDisplay.flag.style.left = evt.clientX+'px';
          this.scoreDisplay.flag.style.top  = parseInt(cPos.y + (canvas.height/2))+'px';
          this.scoreDisplay.pole.style.left = evt.clientX+'px';
          this.scoreDisplay.pole.style.height = cPos.h+'px';
        }
      }
    },
    _showPixelValue: function(scoreDisplay, scores) {
      if(scores) {
        var labels = this.config.labels;

        // don't do alleles when zoomed out
        var scale;
        for(var j=0; this.blocks.length; j++) {
          if(this.blocks[j] != null && this.blocks[j].hasOwnProperty('scale')) {
            scale = this.blocks[j].scale;
            break;
          }
        }
        var map = {};
        for(var j=0; j<scores.length; j++) {
          if(scores[j] == null) continue;
          if(scores[j].feat.data.source === 'counts') {
            map[ scores[j].feat.data.source ] = this.pad(scores[j].score, ' ', 5);
          }
          else {
            map[ scores[j].feat.data.source ] = this.pad(parseFloat(scores[j].score*100).toFixed(2), ' ', 5);
          }
        }

        var countLabel = 'counts';
        if(labels.hasOwnProperty('counts')) countLabel = labels['counts'];

        if(scale < 1) {
          scoreDisplay.innerHTML = countLabel + ': ' + map['counts'];
        }
        else {
          scoreDisplay.style.cssText = scoreDisplay.style.cssText + ' white-space: pre-wrap;'
          scoreDisplay.innerHTML = countLabel + ': ' + map['counts'];

          var keys = Object.keys(map).sort();
          for(var k=0; k<keys.length; k++) {
            var key = keys[k];
            if(key === 'counts') continue;
            var display = key;
            if(labels.hasOwnProperty(key)) display = labels[key];
            scoreDisplay.innerHTML +=  '\n' + this.pad(display, ' ', 5) + ': ' + map[key];
          }
        }
        return true;
      }
      return false;
    },
    pad: function(value, pval, length) {
      return (value.toString().length < length) ? this.pad(pval+value, pval, length):value;
    },
  });
});


js/View/Track/ProportionalWiggle/ProportionalXYPlot.js
define([
  'dojo/_base/declare',
  'dojo/_base/array',
  'dojo/_base/lang',
  'dojo/_base/Color',
  'dojo/on',
  'ProportionalMultiBw/View/Track/ProportionalWiggleBase',
  'JBrowse/View/Track/_YScaleMixin',
  'JBrowse/Util',
  'JBrowse/View/Track/Wiggle/_Scale'
], function(declare, array, lang, Color, on, WiggleBase, YScaleMixin, Util, Scale) {
  return declare([
    WiggleBase, YScaleMixin
  ], {
    _defaultConfig: function() {
      return Util.deepUpdate(lang.clone(this.inherited(arguments)), {
        autoscale: 'local',
        chunkSizeLimit: 500000,
        maxRefFrac: 0.9,
        minRefFrac: 0.0,
        matchRef: true,
        yScalePosition: 'right',
        scale: 'linear',
        style: {
          origin_color: '#888'
        }
      });
    },

    _getScaling: function(viewArgs, successCallback, errorCallback) {
      this._getScalingStats(viewArgs, dojo.hitch(this, function(stats) {
        if (!this.lastScaling || !this.lastScaling.sameStats(stats) || this.trackHeightChanged) {
          var scaling = new Scale(this.config, stats);

          // bump minDisplayed to 0 if it is within 0.5% of it
          if (Math.abs(scaling.min / scaling.max) < 0.005)
            scaling.min = 0;

          // update our track y-scale to reflect it
          this.makeYScale({fixBounds: true, min: scaling.min, max: scaling.max});

          // and finally adjust the scaling to match the ruler's scale rounding
          scaling.min = this.ruler.scaler.bounds.lower;
          scaling.max = this.ruler.scaler.bounds.upper;
          scaling.range = scaling.max - scaling.min;

          this.lastScaling = scaling;
          this.trackHeightChanged = false; // reset flag
        }

        successCallback(this.lastScaling);
      }), errorCallback);
    },

    updateStaticElements: function(coords) {
      this.inherited(arguments);
      this.updateYScaleFromViewDimensions(coords);
    },

    _drawFeatures: function(scale, leftBase, rightBase, block, canvas, pixels, dataScale) {
      var thisB = this;
      var context = canvas.getContext('2d');
      var bConfig = this.browser.config;
      var canvasHeight = canvas.height;

      var lineColor;
      var urlTmp = thisB.config.urlTemplates;
      for (var i = 0; i < urlTmp.length; i++) {
        if (urlTmp[i].name === 'counts') {
          lineColor = urlTmp[i].color;
          break;
        }
      }

      if (scale < 1) {
        var background = new Image();
        background.src = bConfig.baseUrl + 'plugins/ProportionalMultiBw/img/zoom_in.png';
        context.drawImage(background, 0, parseInt((canvasHeight / 2) - 10));
      }

      var refBases = [];
      var matchRef = this.config.matchRef || false;

      var ratio = Util.getResolution(context, bConfig.highResolutionMode);
      var toY = lang.hitch(this, function(val) {
        var norm = 0;
        if (val > 0)
          norm = dataScale.normalize(val);
        return canvasHeight * (1 - norm) / ratio;
      });

      if (scale < 1) {
        var lastXY = {};
        // iterate over pixels (scale pixels per base)
        array.forEach(pixels, function(p, i) {
          var counts = 0;
          // iterate over stores
          array.forEach(p, function(s) {
            if (!s)
              return;
            if (s.feat.get('source') !== 'counts')
              return;
            counts = s.score;
          }, this);

          // line drawing
          var score = toY(counts);
          context.strokeStyle = lineColor;

          context.beginPath();
          if (lastXY.hasOwnProperty('x')) {
            context.moveTo(lastXY.x, lastXY.y);
            context.lineTo(i, score);
            context.stroke();
          }
          lastXY = {
            x: i,
            y: score
          };

        }, this);
      } else {
        thisB.browser.getStore('refseqs', function(refSeqStore) {
          refSeqStore.getReferenceSequence({
            ref: thisB.refSeq.name,
            start: leftBase,
            end: rightBase
          }, dojo.hitch(this, function(seq) {
            refBases = seq.split('');
            // the moved bit starts here
            var maxRefFrac = thisB.config.maxRefFrac;
            var minRefFrac = thisB.config.minRefFrac;

            var colors = {};
            var templates = thisB.config.urlTemplates;
            for (var i = 0; i < templates.length; i++) {
              colors[templates[i].name] = templates[i].color;
            }

            var lastXY = {};
            // iterate over pixels (scale pixels per base)
            array.forEach(pixels, function(p, i) {
              var stack = [];
              var counts = 0;
              // iterate over stores
              array.forEach(p, function(s) {
                if (!s)
                  return;

                var f = s.feat;
                var source = f.get('source');
                var score = s.score;

                if (source === 'counts') {
                  counts = score;
                } else if (score > 0) {
                  var refbase = refBases[f.get('start') - leftBase];
                  if(refbase) {
                    refbase = refbase.toUpperCase()
                  }
                  stack.push({
                    'allele': source,
                    'raw': score,
                    'refBase': refbase
                  });
                }
              }, this);

              var len = stack.length;
              if (stack.length > 0) {
                // test if we should show based on ref base divergence
                var show = false;
                if (matchRef) {
                  for (var j = 0; j < stack.length; j++) {
                    if (stack[j].allele === stack[j].refBase && stack[j].raw >= minRefFrac && stack[j].raw <= maxRefFrac) {
                      show = true;
                      break;
                    }
                  }
                } else {
                  show = true;
                }

                if (show) {
                  // sort them by the raw value desc (or allele asc if equal)
                  stack.sort(function(a, b) {
                    if (a.raw === b.raw) {
                      if (a.allele < b.allele)
                        return -1;
                      if (a.allele > b.allele)
                        return 1;
                    }
                    return b.raw - a.raw;
                  });

                  var height = canvasHeight;
                  for (var k = 0; k < len; k++) {
                    context.fillStyle = colors[stack[k].allele];
                    thisB._fillRectMod(context, i, 0, 1, height);
                    height -= Math.round(canvasHeight * stack[k].raw);
                  }
                }
              }

              // line drawing
              var score = toY(counts);
              context.strokeStyle = lineColor;

              context.beginPath();
              if (lastXY.hasOwnProperty('x')) {
                context.moveTo(lastXY.x, lastXY.y);
                context.lineTo(i, score);
                context.stroke();
              }
              lastXY = {
                x: i,
                y: score
              };

            }, this);
          }),
          // end callback
          function() {},
          // error callback
          dojo.hitch(this, function() {
            console.log(this);
          }));
        });
      }

    }
  });
});
